tree/tree/src/TTree.cxx

```diff
@@ -4438,99 +4441,88 @@ Int_t TTree::Fill()
 
-   if (fAutoFlush != 0 || fAutoSave != 0) {
-      // Is it time to flush or autosave baskets?
-      if (fFlushedBytes == 0) {
-         // Decision can be based initially either on the number of bytes
-         // or the number of entries written.
+   if (gDebug > 0)
+      Info("TTree::Fill", " - A:  %d %lld %lld %lld %lld %lld %lld \n",
+           nbytes, fEntries, fAutoFlush,fAutoSave,GetZipBytes(),fFlushedBytes,fSavedBytes);
+
+   bool autoFlush = fAutoFlush != 0;
+   bool autoSave  = fAutoSave  != 0;
+
+   if (autoFlush || autoSave) {
+      if (fFlushedBytes > 0) {
+         auto flushEntries = fNClusterRange == 0 ? fEntries : fEntries - fClusterRangeEnd[fNClusterRange-1];
+         autoFlush = flushEntries % fAutoFlush == 0;
+         autoSave  = fEntries % fAutoSave == 0;
+      } else {
          Long64_t zipBytes = GetZipBytes();
-         if ((fAutoFlush<0 && zipBytes > -fAutoFlush)  ||
-             (fAutoSave <0 && zipBytes > -fAutoSave )  ||
-             (fAutoFlush>0 && fEntries%TMath::Max((Long64_t)1,fAutoFlush) == 0) ||
-             (fAutoSave >0 && fEntries%TMath::Max((Long64_t)1,fAutoSave)  == 0) ) {
 
-            //First call FlushBasket to make sure that fTotBytes is up to date.
+         autoFlush = fAutoFlush > 0 ? fEntries == fAutoFlush : zipBytes > -fAutoFlush;
+         autoSave  = fAutoSave  > 0 ? fEntries == fAutoSave  : zipBytes > -fAutoSave;
+
+         // First time that AutoFlush/AutoSave are being called
+         if (autoFlush || autoSave) {
             FlushBaskets();
-            OptimizeBaskets(GetTotBytes(),1,"");
-            if (gDebug > 0) Info("TTree::Fill","OptimizeBaskets called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",fEntries,GetZipBytes(),fFlushedBytes);
+            OptimizeBaskets(GetTotBytes(), 1, "");
+            autoFlush = false; // already flushed, do not flush again below
+
+            fAutoFlush = fEntries; // switch to test on number of entries
+
+            if (gDebug > 0)
+               Info("TTree::Fill", "OptimizeBaskets called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",
+                    fEntries, GetZipBytes(), fFlushedBytes);
+
             fFlushedBytes = GetZipBytes();
-            fAutoFlush    = fEntries;  // Use test on entries rather than bytes
-
-            // subsequently in run
-            if (fAutoSave < 0) {
-               // Set fAutoSave to the largest integer multiple of
-               // fAutoFlush events such that fAutoSave*fFlushedBytes
-               // < (minus the input value of fAutoSave)
-               Long64_t totBytes = GetTotBytes();
-               if (zipBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/zipBytes)/fEntries));
-               } else if (totBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/totBytes)/fEntries));
-               } else {
-                  TBufferFile b(TBuffer::kWrite, 10000);
```

Comment:
Indeed, I removed this on the basis that this part of the code is never reached when `fEntries == 0`, and if `fEntries > 0` either `GetZipBytes()` or `GetToBytes()` must return a positive number right after filling baskets, unless there is a serious bug in filling. Is my line of thought wrong? How can you actually have both `GetZipBytes()` and `GetToBytes()` return zero after flushing baskets?